docs/tutorial.rst

```diff
@@ -0,0 +1,144 @@
+
+.. code-block:: python
+
+  net = tensornetwork.TensorNetwork()
+  a = net.add_node(tf.eye(2))
+  b = net.add_node(tf.eye(2))
+  c = net.add_node(tf.eye(2))
+  # Dangling edges are automatically created at node creation. 
+  # We can access them this way.
+  dangling_edge = a.get_edge(1)
+  # This is the same as above
+  dangling_edge = a[1]
+  # Create a standard edge by connecting any two separate nodes together.
+  # We create a new edge by "connecting" two dangling edges.
+  standard_edge = net.connect(a[0], b[0]) 
+  # Create a trace edge by connecting a node to itself.
+  trace_edge = net.connect(c[0], c[1])
+
+Connecting Dangling Edges 
+-------------------------
+One common paradigm in building quantum circuits is to add computation based on edges rather than nodes. That is, it’s very common to want to connect two dangling edges rather than trying to keep track of node axes.
+
+In this example, we have a single qubit quantum circuit where we apply a Hadamard operation several times. We connect the dangling edge of the qubit to the Hadamard operation and return the resulting “output” edge.
+
+.. code-block:: python
+
+  def apply_hadamard(net, edge):
+    hadamard_op = np.array([[1.0, 1.0], [1.0, -1.0]]) / np.sqrt(2.0)
+    hadamard_node = net.add_node(hadamard_op)
+    # Connect the "qubit edge" to the operator "input edge" 
+    net.connect(edge, hadamard_node[0])
+    return hadamard_node[1]  # This is the "output edge".
+
+  # Build the quantum circuit.
+  net = tensornetwork.TensorNetwork()
+  qubit = net.add_node(np.array([1.0, 0.0])) # A "zero state" qubit.
+  qubit_edge = qubit.get_edge(0) # qubit[0] is equivalent.
+  for i in range(5):
+    qubit_edge = apply_hadamard(net, qubit_edge)
+
+Edge Contraction
+----------------
+Contracting an edge is just a simple call. The tensor network API is smart enough to figure out what type of edge was passed and will do the correct computation accordingly.
+
+This example code calculates the dot product of two vectors.
+
+.. code-block:: python
+
+  net = tensornetwork.TensorNetwork()
+  a = net.add_node(tf.ones(2))
+  b = net.add_node(tf.ones(2))
+  edge = net.connect(a[0], b[0])
+  c = net.contract(edge)
+  print(c.tensor.numpy()) # Should print 2.0
+
+
+Optimized Contractions
+----------------------
+During computation, it’s very common for two nodes to have multiple edges connecting each other. If only one of the edges are contracted at a time, then all of the remaining edges become trace edges. This is usually very bad for computation, as the new node will allocate significantly more memory than required. Also, since trace edges only sum the diagonal of the underlying matrix, all of the other values calculated during the first contraction are useless. During contraction, it always more efficent to contract all of these edges at the same time.
```

Comment:
```suggestion
At intermediate states of a computation, it’s very common for two nodes to have multiple edges connecting them. If only one of those edges is contracted, then all of the remaining edges become trace edges. This is usually very inefficient, as the new node will allocate significantly more memory than is ultimately required. Since trace edges only sum the diagonal of the underlying matrix, all of the other values calculated during the first contraction are useless. It is always more efficient to contract all of these edges simultaneously.
```